clients/drptpostprocessor/drptpostprocessor.c

```diff
@@ -0,0 +1,771 @@
+
+        /* Sync decoder to first Packet Stream Boundary (PSB) packet, and then decode
+         * instructions. If there is no PSB packet, the decoder will be synced to the end
+         * of the trace. If error occurs, we will call diagnose_error to print the error
+         * information.
+         * What is PSB packets?
+         * Below answer is quoted from Intel® 64 and IA-32 Architectures Software
+         * Developer’s Manual 32.1.1.1 Packet Summary
+         * “Packet Stream Boundary (PSB) packets: PSB packets act as ‘heartbeats’ that are
+         * generated at regular intervals (e.g., every 4K trace packet bytes). These
+         * packets allow the packet decoder to find the packet boundaries within the
+         * output data stream; a PSB packet should be the first packet that a decoder
+         * looks for when beginning to decode a trace.”
+         */
+        status = pt_insn_sync_forward(decoder->ptdecoder);
+        if (status < 0) {
+            if (status == -pte_eos)
+                break;
+            diagnose_error(decoder->ptdecoder, insn.ip, "sync error", status);
+            break;
+        }
+
+        /* Decode instructions. */
+        for (;;) {
+            /* Handle next status and all pending perf events. */
+            int nextstatus = status;
+            int errcode = 0;
+            while (nextstatus & pts_event_pending) {
+                struct pt_event event;
+
+                nextstatus = pt_insn_event(decoder->ptdecoder, &event, sizeof(event));
+                if (nextstatus < 0) {
+                    break;
+                }
+                /* Ensure the sideband session is synced with the pt insn decode and
+                 * handle sideband events. eg. If there is a context switch sideband
+                 * event, the sideband session will change the Intel PT decoder's memory
+                 * image.
+                 */
+                errcode = sb_event(decoder, &event, options);
+                if (errcode < 0) {
+                    break;
+                }
+            }
+            if (nextstatus < 0) {
+                diagnose_error(decoder->ptdecoder, insn.ip, "handle insn event error",
+                               nextstatus);
+                break;
+            }
+            if (errcode < 0) {
+                diagnose_error(decoder->ptdecoder, insn.ip, "handle sideband event error",
+                               errcode);
+                break;
+            }
+
+            if ((nextstatus & pts_eos) != 0) {
+                break;
+            }
+
+            /* Decode instructions. */
+            status = pt_insn_next(decoder->ptdecoder, &insn, sizeof(insn));
+            if (status < 0) {
+                diagnose_error(decoder->ptdecoder, insn.ip, "decode error", status);
+                break;
+            }
+
+            /*// Todo: converting insn into an instruction-only memref_t
+             * memref_t memref;
```

Comment:
We disallow commented code in our style guide: https://dynamorio.org/page_code_style.html#autotoc_md124